RapidReact/src/main/java/frc/robot/subsystem/ClimberSubsystem.java

```diff
@@ -1,112 +1,311 @@
 import frc.robot.log.*;
-
-import javax.swing.plaf.basic.BasicButtonUI;
+import frc.robot.utils.MotorUtils;
 
 public class ClimberSubsystem extends BitBucketsSubsystem {
 
-  private boolean climberEnabled;
+  // state machine diagram 
+  // https://i.imgur.com/zE1pEXn.png
+  enum ClimbState
+  {
+    Idle,
+    ExtendPartial,
+    ExtendFull,
+    Retract
+  }
+
+  private WPI_TalonSRX climber1;
+  private WPI_TalonSRX climber2;
+  private boolean enabledClimber = true;
+
+  private boolean autoClimb; // is autoclimb enabled
+  private boolean autoClimbPressed = false; // is the autoclimb button currently being pressed
 
-  private WPI_TalonSRX climber;
-  private boolean fixedHookToggleState;
-  private boolean elevatorToggle;
-  private boolean autoClimb;
+  private ClimbState currentClimbState;
   DoubleSolenoid elevatorSolenoid;
-  DoubleSolenoid fixedHookSolenoid;
+  
+  private int fullExtendPosition = 5000; // TODO: change this number
+  private int partialExtendPosition = fullExtendPosition / 2;
+  private int fullRetractPosition = 0; // TODO: change this number
+
+  // https://docs.ctre-phoenix.com/en/stable/ch16_ClosedLoop.html#mechanism-is-finished-command
+  private int climbErrThreshold1;
+  private int climbLoopsToSettle1;
+  private int withinThresholdLoops1;
+
+  private int climbErrThreshold2;
+  private int climbLoopsToSettle2;
+  private int withinThresholdLoops2;
+
+  private boolean climberTilted = false;
 
   private final Changeable<Double> climbOutput = BucketLog.changeable(Put.DOUBLE, "climber/climbOutput", 0.5);
 
   private final Loggable<String> climbState = BucketLog.loggable(Put.STRING, "climber/climbState");
-  private final Loggable<String> hookToggleState = BucketLog.loggable(Put.STRING, "climber/hookState");
-  private final Loggable<String> elevatorToggleState = BucketLog.loggable(Put.STRING, "climber/elevatorState");
+  private final Loggable<String> elevatorTiltedState = BucketLog.loggable(Put.STRING, "climber/elevatorTiltedState");
+  private final Loggable<Double> climberMotorPosition = BucketLog.loggable(Put.DOUBLE, "climber/climberMotorPosition");
 
   public ClimberSubsystem(Config config) {
     super(config);
   }
 
   @Override
   public void init() {
-    climber = new WPI_TalonSRX(config.climberMotor_ID);
+    climber1 = MotorUtils.makeSRX(config.climber.climber1);
+    climber2 = MotorUtils.makeSRX(config.climber.climber2);
+
     if (config.enablePneumatics) {
-      elevatorSolenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH, config.
-      elevatorSolenoid_ID1, config.elevatorSolenoid_ID2);
-      fixedHookSolenoid = new DoubleSolenoid(PneumaticsModuleType.REVPH, config.fixedHookSolenoid_ID1, config.fixedHookSolenoid_ID2);
+      elevatorSolenoid =
+        new DoubleSolenoid(PneumaticsModuleType.REVPH, config.elevatorSolenoid_ID1, config.elevatorSolenoid_ID2);
     }
+
+    currentClimbState = ClimbState.Idle;
+  }
+
+  boolean isClimberAtSetpoint()
+  {
+    climbErrThreshold1 = 10;
```

Comment:
since both sides have the same encoder, you can just use a single variable for the threshold for both of them + a single variable for number of loops. Might consider making the variable in terms of time, not number of loops, if you don't know the frequency that the isClimberAtSetpoint is called.

Reply:
> since both sides have the same encoder

i don't think that's the case - i believe it's two different encoders. they will be the same type, but i dont see how that means i can use a single variable when both encoders can be at different positions?

> if you don't know the frequency that the isClimberAtSetpoint is called

it's every 20 ms, i might change the loop threshold later on depending on how long we want to make sure it's at the right position

Reply:
Yes, that’s correct. Each side has its own encoder. But since they’re of the same type, and controlling the same sort of mechanism, they can both have the same time/ distance thresholds.

Reply:
Ah i see what you mean 👍